iris-mpc-cpu/src/execution/hawk_main.rs

```diff
@@ -454,6 +533,49 @@ impl HawkActor {
         Ok(connect_plan)
     }
 
+    async fn calculate_threshold_a(n_buckets: usize) -> Vec<u32> {
+        let thresholds: Vec<f64> = (0..n_buckets)
+            .map(|i| i as f64 / (n_buckets * 2) as f64)
+            .collect();
+
+        thresholds
+            .iter()
+            .map(|&x| translate_threshold_a(x))
+            .collect()
+    }
+
+    async fn calculate_buckets(
+        n_buckets: usize,
+        session: &mut HawkSession,
+        links: Vec<SortedNeighborhoodV<Aby3Store>>,
+    ) -> Result<VecBuckets> {
+        let translated_thresholds = Self::calculate_threshold_a(n_buckets).await;
+        let mut buckets = vec![];
+        for neighbors in links.iter() {
+            let distances = neighbors.distances_cloned();
+            let bucket_result_shares = compare_threshold_buckets(
+                &mut session.aby3_store.session,
+                translated_thresholds.as_slice(),
+                distances.as_slice(),
+            )
+            .await?;
+
+            let bucket_results =
+                open_ring(&mut session.aby3_store.session, &bucket_result_shares).await?;
+            buckets.push(bucket_results)
```

Comment:
revealing many sub bucket-caclulations here seems to defeat the purpose of aggregating this for privacy a bit.
We could/should instead collect the distances in a large global vec and run a single bucket comparison on it IMO

Reply:
makes sense, will try to adapt to this!